tests/pipeline/test_quarters_estimates.py

```diff
@@ -0,0 +1,955 @@
+                                   4)
+        )
+        for sid, (q1e1, q1e2, q2e1, q2e2) in it:
+            # We're assuming that estimates must come before the relevant
+            # release.
+            if (q1e1 < q1e2 and
+                    q2e1 < q2e2 and
+                    # All estimates are < Q2's event, so just constrain Q1
+                    # estimates.
+                    q1e1 < cls.q1_release_dates[0] and
+                    q1e2 < cls.q1_release_dates[0]):
+                sid_estimates.append(cls.create_estimates_df(q1e1,
+                                                             q1e2,
+                                                             q2e1,
+                                                             q2e2,
+                                                             sid))
+                sid_releases.append(cls.create_releases_df(sid))
+        return pd.concat(sid_estimates +
+                         sid_releases).reset_index(drop=True)
+
+    @classmethod
+    def get_sids(cls):
+        sids = cls.events[SID_FIELD_NAME].unique()
+        # Tack on an extra sid to make sure that sids with no data are
+        # included but have all-null columns.
+        return list(sids) + [max(sids) + 1]
+
+    @classmethod
+    def create_releases_df(cls, sid):
+        # Final release dates never change. The quarters have very tight date
+        # ranges in order to reduce the number of dates we need to iterate
+        # through when testing.
+        return pd.DataFrame({
+            TS_FIELD_NAME: [pd.Timestamp('2015-01-15'),
+                            pd.Timestamp('2015-01-31')],
+            EVENT_DATE_FIELD_NAME: [pd.Timestamp('2015-01-15'),
+                                    pd.Timestamp('2015-01-31')],
+            'estimate': [0.5, 0.8],
+            FISCAL_QUARTER_FIELD_NAME: [1.0, 2.0],
+            FISCAL_YEAR_FIELD_NAME: [2015.0, 2015.0],
+            SID_FIELD_NAME: sid
+        })
+
+    @classmethod
+    def create_estimates_df(cls,
+                            q1e1,
+                            q1e2,
+                            q2e1,
+                            q2e2,
+                            sid):
+        return pd.DataFrame({
+            EVENT_DATE_FIELD_NAME: cls.q1_release_dates + cls.q2_release_dates,
+            'estimate': [.1, .2, .3, .4],
+            FISCAL_QUARTER_FIELD_NAME: [1.0, 1.0, 2.0, 2.0],
+            FISCAL_YEAR_FIELD_NAME: [2015.0, 2015.0, 2015.0, 2015.0],
+            TS_FIELD_NAME: [q1e1, q1e2, q2e1, q2e2],
+            SID_FIELD_NAME: sid,
+        })
+
+    @classmethod
+    def init_class_fixtures(cls):
+        super(WithEstimatesTimeZero, cls).init_class_fixtures()
+
+    def get_expected_estimate(self,
+                              q1_knowledge,
+                              q2_knowledge,
+                              comparable_date):
+        return pd.DataFrame()
+
+    def test_estimates(self):
```

Comment:
I just spent some time fiddling around with the test; tried to do both. This didn't improve the speed, but it did simplify the test a fair amount. If you see any further optimizations happy to incorporate those.